toml/src/main/kotlin/org/rust/toml/completion/crates/local/LocalCargoTomlDependencyCompletionProvider.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.completion.crates.local
+
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.completion.CompletionUtil
+import com.intellij.codeInsight.completion.PrioritizedLookupElement
+import com.intellij.codeInsight.lookup.LookupElement
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.codeInsight.lookup.LookupElementPresentation
+import com.intellij.codeInsight.lookup.LookupElementRenderer
+import com.intellij.icons.AllIcons
+import com.intellij.openapi.components.service
+import org.rust.toml.StringValueInsertionHandler
+import org.rust.toml.completion.CargoNormalizedNamesPrefixMatcher
+import org.rust.toml.completion.TomlKeyValueCompletionProviderBase
+import org.toml.lang.psi.TomlKeyValue
+
+class LocalCargoTomlDependencyCompletionProvider : TomlKeyValueCompletionProviderBase() {
+    override fun completeKey(keyValue: TomlKeyValue, result: CompletionResultSet) {
+        val prefix = CompletionUtil.getOriginalElement(keyValue.key)?.text ?: return
+
+        val indexService = service<CratesLocalIndexService>()
+        if (!indexService.isReady) return
+
+        val crateNames = indexService.getAllCrateNames()
+
+        result.withPrefixMatcher(CargoNormalizedNamesPrefixMatcher(prefix)).addAllElements(crateNames.mapNotNull { crateName ->
+            val crate = indexService.getCrate(crateName)!!
```

Comment:
`!!` looks dangerous here because `getCrate` potentially changes `PersistentHashMap` contents due to `updateIfNeeded` call inside